goodtables/compat.py

```diff
@@ -30,22 +31,22 @@
     basestring = basestring
     numeric_types = (int, long, float)
 
-    def csv_reader(data, dialect=csv.excel, **kwargs):
+    def csv_reader(data, dialect=csv.excel, header_index=0, **kwargs):
         """Read text stream (unicode on Py2.7) as CSV."""
 
-        first_lines = list(islice(data, 5))
-        try:
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
+            dialect = detect_dialect(first_lines)
+            
         def iterenc_utf8(data):
             for line in data:
                 yield line.encode('utf-8')
 
-        iter = chain(first_lines, data)
+        data.seek(0)
```

Comment:
You're changing the behaviour of this function - a behaviour that was once _assumed_ to be the exact opposite.
I want to make sure you checked that no one else in the code base is also assuming that the reader will start from the current position of the stream. 

Coming to think of it, who else in the code base needs the lines before the headers? Why not return an iterator that begins with the header row?